Modules/_asynciomodule.c

```diff
@@ -1601,11 +1601,24 @@ static void
 FutureIter_dealloc(futureiterobject *it)
 {
     PyTypeObject *tp = Py_TYPE(it);
-    asyncio_state *state = get_asyncio_state_by_def((PyObject *)it);
+
+    assert(_PyType_HasFeature((PyTypeObject *)tp, Py_TPFLAGS_HEAPTYPE));
```

Comment:
Maybe add a comment explaining this assert (since `FutureIter` is a heap type, any subclasses must also be).